Source/Csla.Web.Mvc.Shared/ApplicationContextManager.cs

```diff
@@ -0,0 +1,137 @@
+﻿#if NETSTANDARD2_0
+//-----------------------------------------------------------------------
+// <copyright file="ApplicationContextManager.cs" company="Marimer LLC">
+//     Copyright (c) Marimer LLC. All rights reserved.
+//     Website: http://www.lhotka.net/cslanet/
+// </copyright>
+// <summary>Application context manager that uses HttpContextAccessor</summary>
+//-----------------------------------------------------------------------
+using Csla.Core;
+using Microsoft.AspNetCore.Http;
+using System;
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Text;
+
+namespace Csla.Web
+{
+  /// <summary>
+  /// Application context manager that uses HttpContextAccessor when 
+  /// resolving HttpContext to store context values.
+  /// </summary>
+  public class ApplicationContextMananger : IContextManager
+  {
+
+    private const string _localContextName = "Csla.LocalContext";
+    private const string _clientContextName = "Csla.ClientContext";
+    private const string _globalContextName = "Csla.GlobalContext";
+
+    private readonly IServiceProvider _serviceProvider;
+
+    /// <summary>
+    /// Creates an instance of the object, initializing it
+    /// with the required IServiceProvider.
+    /// </summary>
+    /// <param name="serviceProvider">ASP.NET Core IServiceProvider</param>
+    public ApplicationContextMananger(IServiceProvider serviceProvider)
+    {
+      _serviceProvider = serviceProvider;
+    }
+
+    /// <summary>
+    /// Gets the current HttpContext instance.
+    /// </summary>
+    protected virtual HttpContext HttpContext
+    {
+      get
+      {
+        HttpContext result = null;
+        var httpContextAccessor = (IHttpContextAccessor)_serviceProvider.GetService(typeof(IHttpContextAccessor));
+        if (httpContextAccessor != null)
+        {
+          result = httpContextAccessor.HttpContext;
+        }
+        return result;
+      }
+    }
+
+    /// <summary>
+    /// Gets a value indicating whether this
+    /// context manager is valid for use in
+    /// the current environment.
+    /// </summary>
+    public bool IsValid
+    {
+      get { return HttpContext != null; }
+    }
+
+    /// <summary>
+    /// Gets the current principal.
+    /// </summary>
+    public System.Security.Principal.IPrincipal GetUser()
+    {
+      return HttpContext?.User;
+    }
+
+    /// <summary>
+    /// Sets the current principal.
+    /// </summary>
+    /// <param name="principal">Principal object.</param>
+    public void SetUser(System.Security.Principal.IPrincipal principal)
+    {
+      HttpContext.User = (ClaimsPrincipal)principal;
```

Comment:
What do you think about this upcasting? ASP.NET core kind of mandates that your principal class derives from `ClaimsPrincipal` but CSLA doesn't mandate this - and currently the only time you would be aware of the problem is when you hit this cast exception at runtime. Any ideas? My first thought was to use compilation directives to change IContextManager interface so that for asp.net core the method argument is ClaimsPrincipal (is there an IClaimsPrincipal?) rather than IPrincipal? Is that an option?

Reply:
Yeah, I saw that and tried to get rid of the cast. As you note, it is required by ASP.NET Core.

I don't know what to think about that exactly. If we change IContextManager we'd really need to change all uses of IPrincipal through CSLA to use ClaimsPrincipal (or does it have an interface?), and that'd be kind of a Big Deal(tm)...

Reply:
Ok. Well this works for me. If we wanted too, perhaps in the future, we could write our own adaptor to adapt the IPrinciple to a ClaimsPrinciple (assuming it wont cast) on SetUser and then return the original IPrinciple on GetUser. Probably not necessary for the time being.